src/storage/src/source/reclock.rs

```diff
@@ -537,33 +521,24 @@ where
 
     pub async fn mint(
         &mut self,
-        new_source_upper: AntichainRef<'_, FromTime>,
+        binding_ts: IntoTime,
+        mut new_into_upper: Antichain<IntoTime>,
+        new_from_upper: AntichainRef<'_, FromTime>,
     ) -> ReclockBatch<FromTime, IntoTime> {
+        assert!(!new_into_upper.less_equal(&binding_ts));
         // The updates to the remap trace that occured during minting.
         let mut batch = ReclockBatch {
             updates: vec![],
             upper: self.upper.clone(),
         };
 
         while *self.upper == [IntoTime::minimum()]
-            || PartialOrder::less_than(&self.source_upper.frontier(), &new_source_upper)
+            || (PartialOrder::less_equal(&self.source_upper.frontier(), &new_from_upper)
```

Comment:
This has changed from `less_than` to `less_equal` to allow committing a binding in which the upstream frontier hasn't moved but we nevertheless re-observed it. The strict forward progress is guaranteed by the `less_than` which is checked on the into time frontier